pulsar-client-go/pulsar/schema.go

```diff
@@ -0,0 +1,494 @@
+	return json.Unmarshal(data, v)
+}
+
+func (js *JsonSchema) Validate(message []byte) error {
+	return js.UnSerialize(message, nil)
+}
+
+func (js *JsonSchema) GetSchemaInfo() *SchemaInfo {
+	jsonSchema := NewSchemaInfo("Json", js.SchemaInfo.Schema, JSON)
+	return jsonSchema
+}
+
+type ProtoSchema struct {
+	SchemaDefinition
+	SchemaInfo
+}
+
+func NewProtoSchema(codec string) *ProtoSchema {
+	ps := new(ProtoSchema)
+	schema, err := initCodec(codec)
+	if err != nil {
+		log.Fatalf("init codec error:%v", err)
+	}
+	schemaDef := NewSchemaDefinition(schema)
+	ps.SchemaDefinition.SchemaDef = schemaDef.SchemaDef
+	ps.SchemaInfo.Schema = schemaDef.SchemaDef.Schema()
+	ps.SchemaInfo.Type = PROTOBUF
+	return ps
+}
+
+func (ps *ProtoSchema) Serialize(data interface{}) ([]byte, error) {
+	return proto.Marshal(data.(proto.Message))
+}
+
+func (ps *ProtoSchema) UnSerialize(data []byte, v interface{}) error {
+	return proto.Unmarshal(data, v.(proto.Message))
+}
+
+func (ps *ProtoSchema) Validate(message []byte) error {
+	return ps.UnSerialize(message, nil)
+}
+
+func (ps *ProtoSchema) GetSchemaInfo() *SchemaInfo {
+	jsonSchema := NewSchemaInfo("Proto", ps.SchemaDef.Schema(), PROTOBUF)
+	return jsonSchema
+}
+
+type AvroSchema struct {
+	SchemaDefinition
+	SchemaInfo
+}
+
+func NewAvroSchema(codec string) *AvroSchema {
+	as := new(AvroSchema)
+	schema, err := initCodec(codec)
+	if err != nil {
+		log.Fatalf("init codec error:%v", err)
+	}
+	schemaDef := NewSchemaDefinition(schema)
+	as.SchemaDefinition.SchemaDef = schemaDef.SchemaDef
+	as.SchemaInfo.Schema = schemaDef.SchemaDef.Schema()
+	as.SchemaInfo.Type = AVRO
+	return as
+}
+
+func (as *AvroSchema) Serialize(data interface{}) ([]byte, error) {
+	textual, err := json.Marshal(data)
+	if err != nil {
+		log.Errorf("serialize data error:%s", err.Error())
+		return nil, err
+	}
+	native, _, err := as.SchemaDef.NativeFromTextual(textual)
+	if err != nil {
+		log.Errorf("convert native Go form to binary Avro data error:%s", err.Error())
+		return nil, err
+	}
+	return as.SchemaDef.BinaryFromNative(nil, native)
+}
+
+func (as *AvroSchema) UnSerialize(data []byte, v interface{}) error {
+	native, _, err := as.SchemaDef.NativeFromBinary(data)
+	if err != nil {
+		log.Errorf("convert binary Avro data back to native Go form error:%s", err.Error())
+		return err
+	}
+	textual, err := as.SchemaDef.TextualFromNative(nil, native)
+	if err != nil {
+		log.Errorf("convert native Go form to textual Avro data error:%s", err.Error())
+		return err
+	}
+	err = json.Unmarshal(textual, v)
+	if err != nil {
+		log.Errorf("unSerialize textual error:%s", err.Error())
+		return err
+	}
+	return nil
+}
+
+func (as *AvroSchema) Validate(message []byte) error {
+	return as.UnSerialize(message, nil)
+}
+
+func (as *AvroSchema) GetSchemaInfo() *SchemaInfo {
+	jsonSchema := NewSchemaInfo("avro", as.SchemaDef.Schema(), AVRO)
+	return jsonSchema
+}
+
+type StringSchema struct {
+	SchemaDefinition
```

Comment:
why do you need this for `StringSchema`?

Reply:
yes, the `StringSchema` don't need this, will fix it.